### ISSUE TITLE
States of the time field corresponding to ti.state is not clear

### ISSUE THREAD

**Issue by kzosabe** [apache/airflow#38632]
### What do you see as an issue?

There was a problem with the emitting of statistics logs regarding execution time introduced in https://github.com/apache/airflow/pull/30612 .

Problem details:

- https://github.com/apache/airflow/pull/30612#issuecomment-1980985307 
- https://github.com/apache/airflow/pull/34771#discussion_r1542257954

When issuing such statistical indicators of execution time, it is important to first clarify what the corresponding time fields are for each state, and to make recognition and implementation consistent.

The time fields corresponding to the transitions to each state of TIs, as I read the code, are as follows:

```
SCHEDULED: (no time field for this state currently exists)
QUEUED: queued_dttm
RUNNING: start_date
SUCCESS: end_date
FAILED: end_date
SKIPPED: end_date
UPSTREAM_FAILED: end_date
REMOVED: end_date
UP_FOR_RETRY: end_date
UP_FOR_RESCHEDULE: (uncertain)
RESTARTING: (uncertain)
DEFERRED: (uncertain)
```

So as I understood, the matrix of states and TI dates should look like this (corrections welcome):

---

```
SCHEDULED         - queued_dttm: no,  start_date: no,  end_date: no
QUEUED            - queued_dttm: yes, start_date: no,  end_date: no
RUNNING           - queued_dttm: yes, start_date: yes, end_date: no
SUCCESS           - queued_dttm: yes, start_date: yes, end_date: yes(1)
FAILED            - queued_dttm: yes, start_date: yes, end_date: yes(1)
SKIPPED           - queued_dttm: yes, start_date: yes, end_date: yes(1)
UPSTREAM_FAILED   - queued_dttm: yes, start_date: yes, end_date: yes(1)
REMOVED           - queued_dttm: yes, start_date: yes, end_date: yes(1)
UP_FOR_RETRY      - queued_dttm: yes, start_date: yes, end_date: yes(1)
UP_FOR_RESCHEDULE - queued_dttm: yes, start_date: yes, end_date: yes?(2)
RESTARTING        - queued_dttm: no?, start_date: no?, end_date: no?(3)
DEFERRED          - queued_dttm: no?, start_date: no?, end_date: no?(4)
```

(1) https://github.com/apache/airflow/blob/fce3a583348162f655282d032eca654dcb67b497/airflow/models/taskinstance.py#L1894-L1896
and
https://github.com/apache/airflow/blob/33e5d0351d7d09033c57a4a0b851b9e1b50bfb01/airflow/utils/state.py#L160-L168
(2)
https://github.com/apache/airflow/blob/fce3a583348162f655282d032eca654dcb67b497/airflow/models/taskinstance.py#L2829
Is this state being handled this way because the task is interpreted as finished but waiting to be picked up by the scheduler?

(3)
https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/tasks.html#task-instances
One interpretation of the above diagram should look like this.

If RESTARTING and UP_FOR_RETRY are interpreted as pre-scheduled states, then all times should be no.
If we interpret them as finished states like (2), then all times should be yes.
We could also treat these as having indeterminate time field states, but that would probably not be a good idea.

(4)
Undocumented, but commented as `Deferrable operator waiting on a trigger` on this line.
https://github.com/apache/airflow/blob/33e5d0351d7d09033c57a4a0b851b9e1b50bfb01/airflow/utils/state.py#L59
And these lines imply that this state is a pre-scheduled state
https://github.com/apache/airflow/blob/db3181c27bc52ebf0167fb6b2f8181a13e7d02b9/airflow/models/trigger.py#L191-L201

---

### Solving the problem

So it will be great that 

- Fill in the matrix as above by community consensus (and write it into the official document if we can)
- Fix codes that do not follow the matrix above
- Add tests or debug logs that can check for incorrect state and time combinations

Is it worth working on this?

In the first, I tried to tackle the set_state issue mentioned in https://github.com/apache/airflow/pull/34771#discussion_r1542257954 .

https://github.com/apache/airflow/pull/38631

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by uranusjr** [comment]
DEFERRED is the easy one, it is basically the same as RUNNING from the scheduler’s perspective, except it is waiting for the trigger to continue. For a ti to entered the DEFERRED, it must first be in the RUNNING state (because deferral happens in the worker), so it should have set start_date, but not end_date.

UP_FOR_RESCHEDULE is similar, except it’s done by a (synchronous) sensor in reschedule mode. It is also basically the same as RUNNING except waiting for the scheduler to pick it up again. Since signalling a reschedule is done in the worker, the ti should have set start_date. The situation with end_state is… complicated (this uncertainly is a part of the reason why sensors are disfavoured). _Personally_ I would prefer if it does not set end_date (nor re-set start_date when the scheduler restarts the ti), but it is what it is and we can’t change that without breaking compatibility (and when we can break compatibility in Airflow 3, we probably want to remove synchronous sensors altogether anyway). I think UP_FOR_RESCHEDULE should be considered a special state that is also finished and pre-scheduling and keep its current behaviour.

RESTARTING is also similar (it is actually relative new) in that it’s set after the ti have started, so start_date should have been set. It should keep its start_date until the scheduler picks the ti back up, and end_date should remain null. The same applies for UP_FOR_RETRY. (I think your reference on UP_FOR_RETRY’s end_date is wrong? The code it links does not handle UP_FOR_RETRY, from what I can tell.)

**Comment by kzosabe** [comment]
@uranusjr Thanks for the explanation!

Just to confirm, does a DEFFERED task really transition directly to RUNNING?
I may be missing something, but I could not find a code path that directly transitions from DEFFERED to RUNNING.
As far as I have investigated, it seems that DEFFERED tasks are first rolled back to SCHEDULED, as in the following test code.
https://github.com/apache/airflow/blob/db3181c27bc52ebf0167fb6b2f8181a13e7d02b9/tests/models/test_trigger.py#L93-L118

However, whether there is a direct transition or not, your explanation seems to be correct.

As I was checking your comment, I noticed that, at least the presence or absence of start_date is not be determined by the value of ti.state. So the form of matrix I wrote above may be nonsense.

From reading your comment, in my interpretation, the current desired behavior of `start_date` is:

- The datetime when the task is first touched by worker is stored.
- And is normally retained no matter how the state changes thereafter (including changes to pre-running state).
- If there is a action that can be regarded as `restart` , it will be overwritten with the datetime of it.
    - Perhaps we don't want to consider a run from DEFERRED as a `restart`, but we want to consider a run from UP_FOR_RESCHEDULE as a `restart`?

Is this understanding correct?

If this is correct, then at least the changes I've made in https://github.com/apache/airflow/pull/38631 should be reasonable in terms of conforming to this interpretation.
Because in the current implementation, the change from QUEUED to SCHEDULED also sets the end_date, which is also clearly contrary to the above.

However, it seems that changes such as putting None in start_date when reverting to a pre-running state, as described in https://github.com/apache/airflow/pull/38631#discussion_r1545086130 , are wrong and should not be done.


> I think your reference on UP_FOR_RETRY’s end_date is wrong?

According to my referenced code, `ti.set_state(TaskInstanceState.UP_FOR_RETRY)` actually sets current time as ti.end_date (because of `or ti.state == TaskInstanceState.UP_FOR_RETRY`).
However, from reading your commentary, this may also be the wrong behavior.
